vsintegration/src/FSharp.Editor/CodeLens/CodeLens.fs

```diff
@@ -0,0 +1,378 @@
+open Microsoft.CodeAnalysis.Editor.Shared.Extensions
+open Microsoft.CodeAnalysis.Editor.Shared.Utilities
+open Microsoft.CodeAnalysis.Classification
+open Internal.Utilities.StructuredFormat
+open Microsoft.VisualStudio.Text.Tagging
+open System.Collections.Concurrent
+open System.Collections
+open System.Windows.Media.Animation
+
+open Microsoft.VisualStudio.FSharp.Editor.Logging
+
+type CodeLensTag(width, topSpace, baseline, textHeight, bottomSpace, affinity, tag:obj, providerTag:obj) =
+    inherit SpaceNegotiatingAdornmentTag(width, topSpace, baseline, textHeight, bottomSpace, affinity, tag, providerTag)
+
+type internal CodeLens =
+    { TaggedText: Async<(ResizeArray<Layout.TaggedText> * QuickInfoNavigation) option> 
+      mutable Computed: bool }
+
+type internal CodeLensTagger  
+    (
+        workspace: Workspace, 
+        documentId: Lazy<DocumentId>,
+        buffer: ITextBuffer, 
+        checker: FSharpChecker,
+        projectInfoManager: ProjectInfoManager,
+        typeMap: Lazy<ClassificationTypeMap>,
+        gotoDefinitionService: FSharpGoToDefinitionService
+     ) as self =
+    
+    let tagsChanged = new Event<EventHandler<SnapshotSpanEventArgs>,SnapshotSpanEventArgs>()
+    let formatMap = lazy typeMap.Value.ClassificationFormatMapService.GetClassificationFormatMap "tooltip"
+    let codeLensResults = ConcurrentDictionary<string, CodeLens>()
+    let codeLensUIElementCache = ConcurrentDictionary()
+    let currentVisibleLineNumbers = ConcurrentDictionary()
+    let mutable firstTimeChecked = false
+    let mutable bufferChangedCts = new CancellationTokenSource()
+    let mutable layoutChangedCts = new CancellationTokenSource()
+    let mutable view: IWpfTextView option = None
+    let mutable codeLensLayer: IAdornmentLayer option = None
+    let mutable recentFirstVsblLineNmbr, recentLastVsblLineNmbr = 0, 0
+    
+    let layoutTagToFormatting (layoutTag: LayoutTag) =
+        layoutTag
+        |> RoslynHelpers.roslynTag
+        |> ClassificationTags.GetClassificationTypeName
+        |> typeMap.Value.GetClassificationType
+        |> formatMap.Value.GetTextProperties   
+
+    let executeCodeLenseAsync () =  
+        let uiContext = SynchronizationContext.Current
+        asyncMaybe {
+            let! view = view
+            do! Async.Sleep 800 |> liftAsync
+            logInfof "Rechecking code due to buffer edit!"
+            let! document = workspace.CurrentSolution.GetDocument(documentId.Value) |> Option.ofObj
+            let! options = projectInfoManager.TryGetOptionsForEditingDocumentOrProject(document)
+            let! _, _, checkFileResults = checker.ParseAndCheckDocument(document, options, allowStaleResults = true)
+            logInfof "Getting uses of all symbols!"
+            let! symbolUses = checkFileResults.GetAllUsesOfAllSymbolsInFile() |> liftAsync
+            let textSnapshot = view.TextSnapshot.TextBuffer.CurrentSnapshot
+            logInfof "Updating code lens due to buffer edit!"
```

Comment:
even though the message says "code lens" in it using the `[CodeLens]` prefix makes it easier to F3 jump through the log